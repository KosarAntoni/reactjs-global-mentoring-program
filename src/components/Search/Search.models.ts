export interface SearchProps {
  className?: string
  handleSearchSubmit?: (value: string) => void
  defaultValue?: string
}
