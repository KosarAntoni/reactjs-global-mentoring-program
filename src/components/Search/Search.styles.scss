.search {
  display: grid;
  grid-template-columns: 1fr auto;
  grid-template-rows: auto auto;
  gap: 4rem 1.4rem;

  &__heading {
    margin: 0;

    font-size: 4rem;
    font-weight: 300;
    color: var(--white);
  }

  &__input {
    grid-area: 2 / 1;
  }

  &__button {
    grid-area: 2 / 2;
  }
}
