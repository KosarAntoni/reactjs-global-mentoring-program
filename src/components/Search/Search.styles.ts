import styled from '@emotion/styled'

export const StyledSearch = styled.form`
  display: grid;
  grid-template-columns: 1fr auto;
  grid-template-rows: auto auto;
  gap: 4rem 1.4rem;

  & .heading {
    margin: 0;

    font-size: 4rem;
    font-weight: 300;
    color: var(--white);
  }

  & .input {
    grid-area: 2 / 1;
  }

  & .button {
    grid-area: 2 / 2;
  }
`
