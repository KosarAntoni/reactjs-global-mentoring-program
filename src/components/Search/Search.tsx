import React, { FC } from 'react'
import classNames from 'classnames'

import Button from 'components/Button'
import Input from 'components/Input'

import { SearchProps } from './Search.models'

import './Search.styles.scss'

const Search: FC<SearchProps> = ({ className, handleSearchSubmit, defaultValue }) => {
  const handleSubmit = (event: React.SyntheticEvent) => {
    event.preventDefault()
    if (!handleSearchSubmit) return

    const target = event.target as typeof event.target & {
      search: { value: string }
    }
    handleSearchSubmit(target.search?.value)
  }

  return (
    <form className={classNames('search', className)} data-testid='search-node' onSubmit={handleSubmit}>
      <h1 className='search__heading'>FIND YOUR MOVIE</h1>

      <Input
        className='search__input'
        defaultValue={defaultValue || ''}
        name='search'
        placeholder='What do you want to watch?'

      />

      <Button
        className='search__button'
        type='submit'
      >
        Search
      </Button>
    </form>
  )
}

export default Search
