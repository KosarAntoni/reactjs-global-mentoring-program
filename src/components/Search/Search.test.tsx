import React from 'react'
import { fireEvent, render, screen } from '@testing-library/react'

import Search from './Search'

import '@testing-library/jest-dom'

const testFunction = jest.fn()

test('the Search component renders correctly with required and optional props', () => {
  render(<Search className='test-class' defaultValue='input test' handleSearchSubmit={testFunction}/>)

  const node = screen.getByTestId('search-node')

  // the component has been rendered
  expect(node).toBeInTheDocument()

  // generating a snapshot, not for testing purposes but only to see what changes in the HTML structure during CR if the developer updates the snapshot
  expect(node).toMatchSnapshot()

  // has heading
  const headingNode = screen.getByText('FIND YOUR MOVIE')
  expect(headingNode).toBeInTheDocument()

  // has input with correct value
  const inputNode = screen.getByRole('textbox')
  expect(inputNode).toHaveValue('input test')
  expect(inputNode).toHaveAttribute('placeholder', 'What do you want to watch?')

  // after submit right function is called
  const buttonNode = screen.getByRole('button')
  fireEvent.click(buttonNode)
  expect(testFunction).toBeCalled()

  // while submit correct function is called
  fireEvent.submit(node)
  expect(testFunction).toBeCalled()

  // the comonent has correct classes
  expect(node).toHaveClass('search test-class')
})
