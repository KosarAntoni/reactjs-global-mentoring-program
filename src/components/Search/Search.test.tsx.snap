// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Search component renders correctly with required and optional props 1`] = `
<form
  class="search test-class"
  data-testid="search-node"
>
  <h1
    class="search__heading"
  >
    FIND YOUR MOVIE
  </h1>
  <input
    class="input search__input"
    name="search"
    placeholder="What do you want to watch?"
    value="input test"
  />
  <button
    class="button button--solid search__button"
    type="submit"
  >
    Search
  </button>
</form>
`;
